### ISSUE TITLE
--create-bucket-configuration LocationConstraint=eu-west-1

### ISSUE THREAD

**Issue by idkjs** [gpoitch/sammie#16]
How would I pass a the `--create-bucket-configuration LocationConstraint=eu-west-1` flag to sammie?

I am getting this error when running `sammie deploy`
```s
sammie deploy --create-bucket-configuration LocationConstraint=eu-west-3
[sammie] Validating template...
 aws cloudformation validate-template --template-body file://sam.json
[sammie] Template valid ✔︎
[sammie] Creating s3 code bucket (if necessary)...
 aws s3api create-bucket --bucket sam-uploads-086445346371

An error occurred (IllegalLocationConstraintException) when calling the CreateBucket operation: The unspecified location constraint is incompatible for the region specific endpoint this request was sent to.

Error: 255
    at ChildProcess.child.on.code (/Users/prisc_000/.npm-global/lib/node_modules/sammie/src/utils.js:21:47)
    at ChildProcess.emit (events.js:182:13)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
```

[CREATE BUCKET](https://docs.aws.amazon.com/cli/latest/reference/s3api/create-bucket.html) flag should resolve this but can figure out how to pass the flag to sammie. Thank you!

**Comment by gpoitch** [comment]
Perhaps try creating your bucket separately first, since you want a specific configuration.

Then when you deploy, if the `Creating s3 code bucket (if necessary)...` step still fails, we can add another command to explicitly check if the bucket exists first instead of relying on `create-bucket` to do the check.

**Comment by idkjs** [comment]
Thanks @gpoitch. Its not that I wanted a specific config. If i run `sammie init my-app` then `sammie deploy` i get the location error. Here is what i tried following your suggestions. My region is `eu-west-3` if that helps.

`aws s3 mb s3://artifacts-for-sammie-app`
`sammie init sammie-app`
`sammie deploy --s3-bucket artifacts-for-sammie-app`
or just `sammie deploy` or `sammie deploy -p --bucket artifacts-for-sammie-app` 

Either way get below bucket creation error.

Tries to create `--bucket sam-uploads-086445346371` each time.

Full terminal.
```s
PrisBook:sammie-app$ aws s3 mb s3://artifacts-for-sammie-app
make_bucket: artifacts-for-sammie-app
PrisBook:sammie-app$ sammie init sammie-app
[sammie] Getting AWS account id...
 aws sts get-caller-identity
[sammie] Account id: 086445346371 ✔︎
[sammie] Creating project...
[sammie] Created: "sammie-app" - template: sam.json | code: index.js ✔︎
PrisBook:sammie-app$ sammie deploy --s3-bucket artifacts-for-sammie-app
[sammie] Validating template...
 aws cloudformation validate-template --template-body file://sam.json
[sammie] Template valid ✔︎
[sammie] Creating s3 code bucket (if necessary)...
 aws s3api create-bucket --bucket sam-uploads-086445346371

An error occurred (IllegalLocationConstraintException) when calling the CreateBucket operation: Theunspecified location constraint is incompatible for the region specific endpoint this request was sent to.
Error: 255
    at ChildProcess.child.on.code (/Users/prisc_000/.npm-global/lib/node_modules/sammie/src/utils.js:21:47)
    at ChildProcess.emit (events.js:182:13)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
PrisBook:sammie-app$
```

**Comment by gpoitch** [comment]
If you want it to upload code to a specific s3 bucket instead of the one that gets auto-generated for you, you can change that in your `sam.json` under `bucketName` and `sammie deploy` will put it there.

Alternatively, you can do it this way too:
`sammie deploy -p bucketName=artifacts-for-sammie-app`

I'm not clear if you actually wanted to put it in a different bucket other then the auto-generated one.

**Comment by idkjs** [comment]
@gpoitch I did not want to put it in a specific bucket. Putting it in the autogenerated bucket doesnt work with the above commands.

This is the result of running the README.md commands `sammie init my-app` and `sammie deploy`:

```s
PrisBook:SLS$ mkd my-app
PrisBook:my-app$ sammie init my-app
[sammie] Getting AWS account id...
 aws sts get-caller-identity
[sammie] Account id: 086445346371 ✔︎
[sammie] Creating project...
[sammie] Created: "my-app" - template: sam.json | code: index.js ✔︎
PrisBook:my-app$ sammie deploy
[sammie] Validating template...
 aws cloudformation validate-template --template-body file://sam.json
[sammie] Template valid ✔︎
[sammie] Creating s3 code bucket (if necessary)...
 aws s3api create-bucket --bucket sam-uploads-086445346371

An error occurred (IllegalLocationConstraintException) when calling the CreateBucket operation: The unspecified location constraint is incompatible for the region specific endpoint this request was sent to.

Error: 255
    at ChildProcess.child.on.code (/Users/prisc_000/.npm-global/lib/node_modules/sammie/src/utils.js:21:47)
    at ChildProcess.emit (events.js:182:13)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
PrisBook:my-app$
```

So the question is, How do i deal with `IllegalLocationConstraintException`?

I tried passing the `LocationConstraint=eu-west-3` param like so:
`sammie deploy -p LocationConstraint=eu-west-3`

error output:

```s
sammie deploy -p LocationConstraint=eu-west-3
[sammie] Validating template...
 aws cloudformation validate-template --template-body file://sam.json
[sammie] Template valid ✔︎
[sammie] Creating s3 code bucket (if necessary)...
 aws s3api create-bucket --bucket sam-uploads-086445346371

An error occurred (IllegalLocationConstraintException) when calling the CreateBucket operation: The unspecified location constraint is incompatible for the region specific endpoint this request was sent to.

Error: 255
    at ChildProcess.child.on.code (/Users/prisc_000/.npm-global/lib/node_modules/sammie/src/utils.js:21:47)
    at ChildProcess.emit (events.js:182:13)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
```

## Passing bucketName param

The suggested solution returns the same error:

`sammie deploy -p bucketName=artifacts-for-sammie-app`

```s
sammie deploy -p bucketName=artifacts-for-sammie-app
[sammie] Validating template...
 aws cloudformation validate-template --template-body file://sam.json
[sammie] Template valid ✔︎
[sammie] Creating s3 code bucket (if necessary)...
 aws s3api create-bucket --bucket sam-uploads-086445346371

An error occurred (IllegalLocationConstraintException) when calling the CreateBucket operation: The unspecified location constraint is incompatible for the region specific endpoint this request was sent to.

Error: 255
    at ChildProcess.child.on.code (/Users/prisc_000/.npm-global/lib/node_modules/sammie/src/utils.js:21:47)
    at ChildProcess.emit (events.js:182:13)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:237:12)
```

**Comment by gpoitch** [comment]
@idkjs This should be fixed now in v0.6.7. Thanks for the report.

The auto-created bucket should now work with all regions.  I tested it using eu-west-3.  Regions outside us-east-1 needed additional info and I hadn't noticed because I developed using us-east-1.

The `-p bucketName=x` doesn't work - it should just be set in your sam.json if you ever want to change the bucket.  I added that to the readme.
